subsys/sd/mmc.c

```diff
@@ -34,14 +50,491 @@ inline int mmc_ioctl(struct sd_card *card, uint8_t cmd, void *buf)
+	cmd.response_type = SD_RSP_TYPE_R1b;
+	cmd.timeout_ms = CONFIG_SD_CMD_TIMEOUT;
+	ret = sdhc_request(card->sdhc, &cmd, NULL);
+	if (ret) {
+		LOG_DBG("Error setting bus timing: %d", ret);
+		return ret;
+	}
+	sdmmc_wait_ready(card);
+	card->bus_io.clock = 52000000;
+        card->bus_io.timing = SDHC_TIMING_HS;
+	ret = sdhc_set_io(card->sdhc, &card->bus_io);
+        if (ret) {
+                return ret;
+        }
+
+	return 0;
+}
+
+static int mmc_bus_test(struct sd_card *card)
+{
+	/*Currently WIP*/
+	return 0;
+
+	int ret;
+	struct sdhc_command cmd = {0};
+	struct sdhc_data data = {0};
+	uint8_t* bus_test_data;
+	uint8_t* bus_test_mask;
+	uint8_t bus_8bit_test_data[8] = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xAA, 0x55};
+	uint8_t bus_4bit_test_data[4] = {0x00, 0x00, 0x00, 0x5A};
+	uint8_t bus_1bit_test_data[1] = {0x80};
+	uint8_t bus_8bit_test_mask[8] = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xFF};
+	uint8_t bus_4bit_test_mask[4] = {0x00, 0x00, 0x00, 0xFF};
+	uint8_t bus_1bit_test_mask[1] = {0xC0};
 
-	ret = 0;
+	switch (card->bus_io.bus_width) {
+		case SDHC_BUS_WIDTH8BIT:
+			data.blocks = 8;
+			bus_test_data = bus_8bit_test_data;
+			bus_test_mask = bus_8bit_test_mask;
+			break;
+		case SDHC_BUS_WIDTH4BIT:
+			data.blocks = 4;
+			bus_test_data = bus_4bit_test_data;
+			bus_test_mask = bus_4bit_test_mask;
+			break;
+		case SDHC_BUS_WIDTH1BIT:
+			data.blocks = 1;
+			bus_test_data = bus_1bit_test_data;
+			bus_test_mask = bus_1bit_test_mask;
+			break;
+		default:
+			return -ENOTSUP;
+	}
+
+	cmd.opcode = MMC_SEND_BUS_TEST;
+	cmd.response_type = SD_RSP_TYPE_R1;
+	cmd.timeout_ms = CONFIG_SD_CMD_TIMEOUT;
+	data.block_size = 1;
+	data.data = bus_test_data;
+	data.timeout_ms = CONFIG_SD_DATA_TIMEOUT;
+	ret = sdhc_request(card->sdhc, &cmd, NULL);
+	if (ret) {
+		return ret;
+	}
+
+	cmd.opcode = MMC_CHECK_BUS_TEST;
+        cmd.response_type = SD_RSP_TYPE_R1;
+        cmd.timeout_ms = CONFIG_SD_CMD_TIMEOUT;
+	data.data = card->card_buffer;
+	ret = sdhc_request(card->sdhc, &cmd, &data);
+        if (ret) {
+                return ret;
+        }
+	for (int i = 0; i < data.blocks; i++) {
+		if ((((uint8_t*)data.data)[i]) & (bus_test_mask[i]))
+		{
+			LOG_ERR("Bus test failed");
+			return -EIO;
+		}
+	}
 
 	return ret;
 }
+
+static int mmc_set_bus_width(struct sd_card *card)
+{
+	int ret;
+	struct sdhc_command cmd = {0};
+
+	if (card->host_props.host_caps.bus_8_bit_support) {
+		cmd.opcode = SD_SWITCH;
+		cmd.arg = MMC_SWITCH_8_BIT_BUS_ARG;
+		cmd.response_type = SD_RSP_TYPE_R1b;
+		cmd.timeout_ms = CONFIG_SD_CMD_TIMEOUT;
+
+		ret = sdhc_request(card->sdhc, &cmd, NULL);
+		sdmmc_wait_ready(card);
+		if (ret) {
+			LOG_DBG("Setting card data bus width failed: %d", ret);
+			return ret;
+		}
+
+		card->bus_io.bus_width = SDHC_BUS_WIDTH8BIT;
+		ret = sdhc_set_io(card->sdhc, &card->bus_io);
+		if (ret) {
+			LOG_DBG("Setting SDHC data bus width failed: %d", ret);
+			return ret;
+		}
+
+		//ret = mmc_bus_test_8_bit(card);
```

Comment:
nit: should this be uncommented?

Reply:
actually it should be deleted because it doesn't exist